rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+	repeated string dependencies = 6;
+
+	// required_versions is a repeated string of service version	// requirements. e.g. "clouddriver >= 5.33.0"
+	repeated string required_versions = 7;
+
+	// A list of extension points that the plugin implements.
+	// e.g. "pipelineStage"
+	repeated string provided_capabilities = 8;
+
+	// Whether or not the plugin is enabled. An unset value 
+	// should be considered as true.
+	boolean enabled = 9;
+}
+```
+
+### Protobuf as the Remote Contract
+As a departure from PF4J, plugin contracts will be defined as Protobuf messages (not services), so long as they are _remote capable_. If a plugin is in-process only, and there’s no expectation that it becomes available for remote plugins, it does not need to define its contract as protobuf.
+
+However, for any remotely-capable extension, its contract must be defined as protobufs, rather than exposed as simple POJOs or interfaces. A key distinction is that we’re not saying gRPC: A remote transport may be Amazon SQS or Redis PubSub, not necessarily gRPC.
+
+### Technology Compatibility Kits
+_alias: TCKs_
+
+A convention of writing TCKs for each Extension Point will be established. The concept of a TCK is borrowed from the JSR world, where compatibility tests are written to ensure that various JREs must implement to assert correct behavior of various features. For Spinnaker plugins, TCKs can be written for two different consumers: Extension Point providers and Plugin developers.
+
+TCKs satisfy asserting correct behavior of an extension point and its associated plugin(s). For any extension point, a TCK will be written that asserts even pathological plugins behave in an expected way while integrating with the extension point. For a plugin, a TCK will be provided that a plugin developer can use to assert correctness without necessarily needing to run other Spinnaker services.
+
+This goes further than the convention of using Protobuf to define the interfaces of a particular plugin, where Protobufs only define the interface, whereas a TCK will assert behavior that lies beneath the interface. Combining a TCK and adherence to the interface, contracts will be fully asserted between a plugin and its extension point(s).
+
+### Up/Downgrading Plugins
+This section applies to federated development scenarios, where the operator is not necessarily the same as the plugin developers; where the release cadence of a plugin does not adhere to the release cadence of the core Spinnaker services.
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+When a service starts up, PF4J will scan a designated locally-accessible directory for plugins to install (these plugins may be downloaded at runtime or baked into the image at built-time). Once an application has started, it will notify Front50 of the plugins that are installed, and get a response dictating which plugins are to be enabled.
```

Comment:
can you elaborate on why this order of operations?
e.g. I would expect, given that Front50 is source of truth, that we call Front50 first, then try to load the enabled plugins. This way the flow is the same of downloaded/not downloaded plugins

Reply:
Yeah, I can get behind this flow. Thank you both.

Reply:
Ah this is new news to me! I was under the impression we were trying to move away from services downloading plugin resources, is that not the case anymore? 

From my experience, most/all people running spinnaker are also running a secure artifact store, eg artifactory. I would expect plugin resources would be stored there and referenced from Front50, so having Front50 be the source of truth for where plugins are loaded is good with me.

Reply:
Having a download story seems a necessity if we want to decouple the release cycles of the Spinnaker services and their plugins. We're not yet going to address runtime updates or anything of the sort, so it's not going to be as flexible as a remote plugin, but it's something.

Front50 won't be the artifact store, but just the storage that knows what plugins are installed & enabled. Artifactory, some blob store, or whatever will still be needed to actually download the artifacts from.

Reply:
Do we need to include anything on expectations for supported download protocols and authentication mechanisms?

Reply:
I added a bit about checkums. I don't intend to go deep on the artifact store or the supported protocols. I think investigating the possibility of using `kork-artifacts` might be a decent idea. We definitely don't want to reinvent the artifact repository story - for example, we would likely use Artifactory here at Netflix.